File: internal/mapping/variable_set.go
package mapping

import (
	"github.com/distr-sh/distr/api"
	"github.com/distr-sh/distr/internal/types"
)

func VariableSetToAPI(variableSet types.VariableSet) api.VariableSet {
	return api.VariableSet{
		ID:             variableSet.ID,
		CreatedAt:      variableSet.CreatedAt,
		UpdatedAt:      variableSet.UpdatedAt,
		Name:           variableSet.Name,
		Description:    variableSet.Description,
		SortOrder:      variableSet.SortOrder,
		ApplicationIDs: variableSet.ApplicationIDs,
		Variables:      List(variableSet.Variables, VariableToAPI),
	}
}

func VariableToAPI(variable types.Variable) api.Variable {
	return api.Variable{
		ID:            variable.ID,
		CreatedAt:     variable.CreatedAt,
		UpdatedAt:     variable.UpdatedAt,
		Key:           variable.Key,
		Description:   variable.Description,
		Type:          api.VariableType(variable.Type),
		IsRequired:    variable.IsRequired,
		DefaultValue:  variable.DefaultValue,
		ReferenceID:   variable.ReferenceID,
		ReferenceName: variable.ReferenceName,
		ScopedValues:  List(variable.ScopedValues, VariableScopedValueToAPI),
	}
}

func VariableScopedValueToAPI(scopedValue types.VariableScopedValue) api.VariableScopedValue {
	return api.VariableScopedValue{
		ID:            scopedValue.ID,
		CreatedAt:     scopedValue.CreatedAt,
		UpdatedAt:     scopedValue.UpdatedAt,
		Scope:         VariableScopeToAPI(scopedValue.Scope),
		SortOrder:     scopedValue.SortOrder,
		Value:         scopedValue.Value,
		ReferenceID:   scopedValue.ReferenceID,
		ReferenceName: scopedValue.ReferenceName,
	}
}

func VariableScopeToAPI(scope types.VariableScope) api.VariableScope {
	return api.VariableScope{
		CustomerOrganizationID: scope.CustomerOrganizationID,
		EnvironmentID:          scope.EnvironmentID,
		ChannelID:              scope.ChannelID,
		DeploymentTargetID:     scope.DeploymentTargetID,
		ApplicationID:          scope.ApplicationID,
		TargetTag:              scope.TargetTag,
		ProcessStepKey:         scope.ProcessStepKey,
	}
}

func ResolvedVariableToAPI(variable types.ResolvedVariable) api.ResolvedVariable {
	return api.ResolvedVariable{
		VariableSetID: variable.VariableSetID,
		VariableID:    variable.VariableID,
		Key:           variable.Key,
		Type:          api.VariableType(variable.Type),
		IsRequired:    variable.IsRequired,
		Status:        string(variable.Status),
		Source:        string(variable.Source),
		Value:         variable.Value,
		ReferenceID:   variable.ReferenceID,
		ReferenceName: variable.ReferenceName,
		Redacted:      variable.Redacted,
		Trace:         List(variable.Trace, VariableResolutionTraceEntryToAPI),
	}
}

func VariableResolutionTraceEntryToAPI(entry types.VariableResolutionTraceEntry) api.VariableResolutionTraceEntry {
	return api.VariableResolutionTraceEntry{
		Source:   string(entry.Source),
		Scope:    VariableScopeToAPI(entry.Scope),
		Selected: entry.Selected,
		Reason:   entry.Reason,
	}
}

File: frontend/ui/src/app/types/variable-set.ts
export type VariableType =
  | 'string'
  | 'number'
  | 'boolean'
  | 'json'
  | 'secret_reference'
  | 'account_reference'
  | 'certificate_reference';

export interface VariableSet {
  id: string;
  createdAt: string;
  updatedAt: string;
  name: string;
  description: string;
  sortOrder: number;
  applicationIds: string[];
  variables: Variable[];
}

export interface Variable {
  id: string;
  createdAt: string;
  updatedAt: string;
  key: string;
  description: string;
  type: VariableType;
  isRequired: boolean;
  defaultValue?: unknown;
  referenceId?: string;
  referenceName?: string;
  scopedValues?: VariableScopedValue[];
}

export interface CreateUpdateVariableSetRequest {
  name: string;
  description: string;
  sortOrder: number;
  applicationIds: string[];
  variables: VariableRequest[];
}

export interface VariableRequest {
  key: string;
  description: string;
  type: VariableType;
  isRequired: boolean;
  defaultValue?: unknown;
  referenceId?: string;
  referenceName?: string;
  scopedValues?: VariableScopedValueRequest[];
}

export interface VariableScope {
  customerOrganizationId?: string;
  environmentId?: string;
  channelId?: string;
  deploymentTargetId?: string;
  applicationId?: string;
  targetTag?: string;
  processStepKey?: string;
}

export interface VariableScopedValue {
  id: string;
  createdAt: string;
  updatedAt: string;
  scope: VariableScope;
  sortOrder: number;
  value?: unknown;
  referenceId?: string;
  referenceName?: string;
}

export interface VariableScopedValueRequest {
  scope: VariableScope;
  sortOrder: number;
  value?: unknown;
  referenceId?: string;
  referenceName?: string;
}

export interface ResolveVariablesPreviewRequest {
  variableSetIds: string[];
  scope: VariableResolutionScope;
  promptedValues?: VariablePromptedValue[];
}

export interface VariableResolutionScope {
  customerOrganizationId?: string;
  environmentId?: string;
  channelId?: string;
  deploymentTargetId?: string;
  applicationId?: string;
  targetTags?: string[];
  processStepKey?: string;
}

export interface VariablePromptedValue {
  key: string;
  value?: unknown;
  referenceId?: string;
  referenceName?: string;
}

export interface ResolvedVariable {
  variableSetId: string;
  variableId: string;
  key: string;
  type: VariableType;
  isRequired: boolean;
  status: 'resolved' | 'unresolved';
  source: string;
  value?: unknown;
  referenceId?: string;
  referenceName?: string;
  redacted: boolean;
  trace: VariableResolutionTraceEntry[];
}

export interface VariableResolutionTraceEntry {
  source: string;
  scope: VariableScope;
  selected: boolean;
  reason: string;
}
